openapi3/schema_merge.go

```diff
@@ -0,0 +1,789 @@
+	for i, val := range values {
+		uintValues[i] = uint64(val)
+	}
+	lcmValue := uintValues[0]
+	for _, v := range uintValues {
+		lcmValue = lcm(lcmValue, v)
+	}
+	schema.MultipleOf = Float64Ptr(float64(lcmValue) / factor)
+	return schema
+}
+
+func getPropFieldsToMerge(collection *SchemaCollection) []string {
+	properties := [][]string{}
+	for i, schema := range collection.Properties {
+		additionalProps := collection.AdditionalProperties[i].Has
+		if additionalProps != nil && !*additionalProps {
+			keys := []string{}
+			for key := range schema {
+				keys = append(keys, key)
+			}
+			properties = append(properties, keys)
+		}
+	}
+	if len(properties) > 0 {
+		return findIntersection(properties...)
+	}
+	keys := []string{}
+	for _, schema := range collection.Properties {
+		for key := range schema {
+			keys = append(keys, key)
+		}
+	}
+	return keys
+}
+
+func resolveProperties(schema *Schema, collection *SchemaCollection) (*Schema, error) {
+	keys := getPropFieldsToMerge(collection)
+	allRefs := map[string][]*Schema{}              //naming
+	for _, schema := range collection.Properties { //naming
+		for name, schemaRef := range schema {
+			if containsString(keys, name) {
+				allRefs[name] = append(allRefs[name], schemaRef.Value)
+			}
+		}
+	}
+
+	result := make(Schemas)
+	for name, schemas := range allRefs {
+		merged, err := mergeFields(schemas)
+		if err != nil {
+			schema.Properties = nil
+			return schema, err
+		}
+		ref := SchemaRef{
+			Value: merged,
+		}
+		result[name] = &ref
+	}
+	if len(result) == 0 {
+		result = nil
+	}
+
+	schema.Properties = result
+	return schema, nil
+}
+
+func resolveAdditionalProperties(schema *Schema, collection *SchemaCollection) (*Schema, error) {
+	additionalProperties := &AdditionalProperties{
+		Has:    nil,
+		Schema: nil,
+	}
+
+	additionalSchemas := []*Schema{}
+	for _, ap := range collection.AdditionalProperties {
+		if ap.Has != nil && !*ap.Has {
+			hasValue := false
+			additionalProperties.Has = &hasValue
+			schema.AdditionalProperties = *additionalProperties
+			return schema, nil
+		}
+		if ap.Schema != nil && ap.Schema.Value != nil {
+			additionalSchemas = append(additionalSchemas, ap.Schema.Value)
+		}
+	}
+
+	if len(additionalSchemas) > 0 {
+		result, err := mergeFields(additionalSchemas)
+		if err != nil {
+			return schema, err
+		}
+		additionalProperties.Schema = &SchemaRef{
+			Value: result,
+		}
+	}
+
+	schema.AdditionalProperties = *additionalProperties
+	return schema, nil
+}
+
+func resolveEnum(values [][]interface{}) []interface{} {
+	var nonEmptyEnum [][]interface{}
+	for _, enum := range values {
+		if len(enum) > 0 {
+			nonEmptyEnum = append(nonEmptyEnum, enum)
+		}
+	}
+	return findIntersectionOfArrays(nonEmptyEnum)
+}
+
+func resolvePattern(values []string) string {
+	var pattern strings.Builder
+	for _, p := range values {
+		if len(p) > 0 {
+			if !isPatternResolved(p) {
+				pattern.WriteString(fmt.Sprintf("(?=%s)", p))
```

Comment:
@tcdsv can you explain how this combines the patterns?

Reply:
we use positive lookahead (?=<pattern>) to specify that regex patterns must follow eachother, this enabled to combine multiple patterns.

The part you're talking about helps us avoid double-wrapping a pattern with "(?=..)"
This suggests a refactor for making that intent clearer.
Also, I've noticed that we don't have any validation tests for the pattern, I will add it.